bug: `localStorage` methods give leaking resource errors in `Deno.test`

Basically, if you use `localStorage.getItem` or `localStorage.setItem` within a `Deno.test` callback (async or not), you will get errors for leaking resources.

## Example

```ts
Deno.test("Getting Keys", () => {
  localStorage.getItem("hello");
});

Deno.test("Setting Keys", () => {
  localStorage.setItem("hello", "foo");
});
```

```sh
$ deno test --location "http://deno.test"
```

## What Happens

It gives this error:

```
Check file:///C:/Projects/Deno/safe/mod_test.ts
running 2 tests from file:///C:/Projects/Deno/safe/mod_test.ts
test Getting Keys ... FAILED (9ms)
test Setting Keys ... ok (28ms)

failures:

Getting Keys
AssertionError: Test case is leaking resources.
Before: {
  "0": "stdin",
  "1": "stdout",
  "2": "stderr"
}
After: {
  "0": "stdin",
  "1": "stdout",
  "2": "stderr",
  "3": "webStorage"
}

Make sure to close all open resource handles returned from Deno APIs before
finishing test case.
    at assert (deno:runtime/js/06_util.js:33:13)
    at resourceSanitizer (deno:runtime/js/40_testing.js:69:7)
    at async exitSanitizer (deno:runtime/js/40_testing.js:85:9)
    at async runTest (deno:runtime/js/40_testing.js:199:7)
    at async Object.runTests (deno:runtime/js/40_testing.js:244:7)
    at async file:///C:/projects/deno/safe/$deno$test.js:1:1

failures:

        Getting Keys

test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out (88ms)
```

## Expect behavior

It should not be throwing errors.

## Notes

If you actually look closely at the stdout, then you can see that it says that the "Setting Keys" test passes.  It seems that you will only get the resource leak error for the first `localStorage` call.  You can see this by changing the code to the following:

```ts
Deno.test("Getting Keys", () => { // FAIL
  localStorage.getItem("hello");
});

Deno.test("Setting Keys #1", () => { // PASS
  localStorage.setItem("hello", "foo");
});

Deno.test("Setting Keys #2", () => { // PASS
  localStorage.setItem("hello", "foo3");
});

Deno.test("foo", () => { // FAIL
  throw Error("foo");
});
```

--- a/extensions/webstorage/01_webstorage.js
+++ b/extensions/webstorage/01_webstorage.js
@@ -1,9 +1,5 @@
 // Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
 
-
-
-
-
 ((window) => {
   const core = window.Deno.core;
   const webidl = window.__bootstrap.webidl;

--- a/extensions/webstorage/01_webstorage.js
+++ b/extensions/webstorage/01_webstorage.js
@@ -1,3 +1,9 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+
+
+
+
 ((window) => {
   const core = window.Deno.core;
   const webidl = window.__bootstrap.webidl;

--- a/extensions/webstorage/01_webstorage.js
+++ b/extensions/webstorage/01_webstorage.js
@@ -2,10 +2,10 @@
   const core = window.Deno.core;
   const webidl = window.__bootstrap.webidl;
 
-  const _rid = Symbol("[[rid]]");
+  const _persistent = Symbol("[[persistent]]");
 
   class Storage {
-    [_rid];
+    [_persistent];
 
     constructor() {
       webidl.illegalConstructor();
@@ -13,7 +13,7 @@
 
     get length() {
       webidl.assertBranded(this, Storage);
-      return core.opSync("op_webstorage_length", this[_rid]);
+      return core.opSync("op_webstorage_length", this[_persistent]);
     }
 
     key(index) {
@@ -25,10 +25,7 @@
         context: "Argument 1",
       });
 
-      return core.opSync("op_webstorage_key", {
-        rid: this[_rid],
-        index,
-      });
+      return core.opSync("op_webstorage_key", index, this[_persistent]);
     }
 
     setItem(key, value) {
@@ -45,10 +42,9 @@
       });
 
       core.opSync("op_webstorage_set", {
-        rid: this[_rid],
         keyName: key,
         keyValue: value,
-      });
+      }, this[_persistent]);
     }
 
     getItem(key) {
@@ -60,10 +56,7 @@
         context: "Argument 1",
       });
 
-      return core.opSync("op_webstorage_get", {
-        rid: this[_rid],
-        keyName: key,
-      });
+      return core.opSync("op_webstorage_get", key, this[_persistent]);
     }
 
     removeItem(key) {
@@ -75,25 +68,20 @@
         context: "Argument 1",
       });
 
-      core.opSync("op_webstorage_remove", {
-        rid: this[_rid],
-        keyName: key,
-      });
+      core.opSync("op_webstorage_remove", key, this[_persistent]);
     }
 
     clear() {
       webidl.assertBranded(this, Storage);
-      core.opSync("op_webstorage_clear", this[_rid]);
+      core.opSync("op_webstorage_clear", this[_persistent]);
     }
   }
 
   function createStorage(persistent) {
     if (persistent) window.location;
 
-    const rid = core.opSync("op_webstorage_open", persistent);
-
     const storage = webidl.createBranded(Storage);
-    storage[_rid] = rid;
+    storage[_persistent] = persistent;
 
     const proxy = new Proxy(storage, {
       deleteProperty(target, key) {
@@ -135,7 +123,7 @@
         return (typeof target.getItem(p)) === "string";
       },
       ownKeys() {
-        return core.opSync("op_webstorage_iterate_keys", rid);
+        return core.opSync("op_webstorage_iterate_keys", persistent);
       },
       getOwnPropertyDescriptor(target, key) {
         if (arguments.length === 1) {

--- a/extensions/webstorage/Cargo.toml
+++ b/extensions/webstorage/Cargo.toml
@@ -1,4 +1,4 @@
-# Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+# Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
 
 [package]
 name = "deno_webstorage"

--- a/extensions/webstorage/lib.rs
+++ b/extensions/webstorage/lib.rs
@@ -46,32 +46,43 @@ pub fn get_declaration() -> PathBuf {
   PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("lib.deno_webstorage.d.ts")
 }
 
+struct LocalStorage(Connection);
+struct SessionStorage(Connection);
+
 fn get_webstorage(
   state: &mut OpState,
   persistent: bool,
 ) -> Result<&Connection, AnyError> {
-  if state.try_borrow::<Connection>().is_none() {
-    let conn = if persistent {
+  let conn = if persistent {
+    if state.try_borrow::<LocalStorage>().is_none() {
       let path = state.try_borrow::<LocationDataDir>().ok_or_else(|| {
         DomExceptionNotSupportedError::new(
           "LocalStorage is not supported in this context.",
         )
       })?;
       std::fs::create_dir_all(&path.0)?;
-      Connection::open(path.0.join("local_storage"))?
-    } else {
-      Connection::open_in_memory()?
-    };
+      let conn = Connection::open(path.0.join("local_storage"))?;
+      state.put(LocalStorage(conn));
+    }
 
-    conn.execute(
-      "CREATE TABLE IF NOT EXISTS data (key VARCHAR UNIQUE, value VARCHAR)",
-      params![],
-    )?;
 
-    state.put(conn);
-  }
 
-  Ok(state.borrow::<Connection>())
+    &state.borrow::<LocalStorage>().0
+  } else {
+    if state.try_borrow::<SessionStorage>().is_none() {
+      let conn = Connection::open_in_memory()?;
+      state.put(SessionStorage(conn));
+    }
+
+    &state.borrow::<SessionStorage>().0
+  };
+
+  conn.execute(
+    "CREATE TABLE IF NOT EXISTS data (key VARCHAR UNIQUE, value VARCHAR)",
+    params![],
+  )?;
+
+  Ok(conn)
 }
 
 pub fn op_webstorage_length(

--- a/extensions/webstorage/lib.rs
+++ b/extensions/webstorage/lib.rs
@@ -62,6 +62,11 @@ fn get_webstorage(
       })?;
       std::fs::create_dir_all(&path.0)?;
       let conn = Connection::open(path.0.join("local_storage"))?;
+      conn.execute(
+        "CREATE TABLE IF NOT EXISTS data (key VARCHAR UNIQUE, value VARCHAR)",
+        params![],
+      )?;
+
       state.put(LocalStorage(conn));
     }
 
@@ -69,16 +74,16 @@ fn get_webstorage(
   } else {
     if state.try_borrow::<SessionStorage>().is_none() {
       let conn = Connection::open_in_memory()?;
+      conn.execute(
+        "CREATE TABLE data (key VARCHAR UNIQUE, value VARCHAR)",
+        params![],
+      )?;
+
       state.put(SessionStorage(conn));
     }
 
     &state.borrow::<SessionStorage>().0
   };
-
-  conn.execute(
-    "CREATE TABLE IF NOT EXISTS data (key VARCHAR UNIQUE, value VARCHAR)",
-    params![],
-  )?;
 
   Ok(conn)
 }

--- a/extensions/webstorage/lib.rs
+++ b/extensions/webstorage/lib.rs
@@ -65,8 +65,6 @@ fn get_webstorage(
       state.put(LocalStorage(conn));
     }
 
-
-
     &state.borrow::<LocalStorage>().0
   } else {
     if state.try_borrow::<SessionStorage>().is_none() {

--- a/extensions/webstorage/lib.rs
+++ b/extensions/webstorage/lib.rs
@@ -46,10 +46,10 @@ pub fn get_declaration() -> PathBuf {
   PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("lib.deno_webstorage.d.ts")
 }
 
-fn get_webstorage<'a>(
-  state: &'a mut OpState,
+fn get_webstorage(
+  state: &mut OpState,
   persistent: bool,
-) -> Result<&'a Connection, AnyError> {
+) -> Result<&Connection, AnyError> {
   if state.try_borrow::<Connection>().is_none() {
     let conn = if persistent {
       let path = state.try_borrow::<LocationDataDir>().ok_or_else(|| {

--- a/extensions/webstorage/lib.rs
+++ b/extensions/webstorage/lib.rs
@@ -1,18 +1,14 @@
 // Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
 
-use deno_core::error::bad_resource_id;
 use deno_core::error::AnyError;
 use deno_core::include_js_files;
 use deno_core::op_sync;
 use deno_core::Extension;
 use deno_core::OpState;
-use deno_core::Resource;
-use deno_core::ZeroCopyBuf;
 use rusqlite::params;
 use rusqlite::Connection;
 use rusqlite::OptionalExtension;
 use serde::Deserialize;
-use std::borrow::Cow;
 use std::fmt;
 use std::path::PathBuf;
 
@@ -26,7 +22,6 @@ pub fn init(location_data_dir: Option<PathBuf>) -> Extension {
       "01_webstorage.js",
     ))
     .ops(vec![
-      ("op_webstorage_open", op_sync(op_webstorage_open)),
       ("op_webstorage_length", op_sync(op_webstorage_length)),
       ("op_webstorage_key", op_sync(op_webstorage_key)),
       ("op_webstorage_set", op_sync(op_webstorage_set)),
@@ -51,82 +46,59 @@ pub fn get_declaration() -> PathBuf {
   PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("lib.deno_webstorage.d.ts")
 }
 
-struct WebStorageConnectionResource(Connection);
-
-impl Resource for WebStorageConnectionResource {
-  fn name(&self) -> Cow<str> {
-    "webStorage".into()
-  }
-}
-
-pub fn op_webstorage_open(
-  state: &mut OpState,
+fn get_webstorage<'a>(
+  state: &'a mut OpState,
   persistent: bool,
-  _zero_copy: Option<ZeroCopyBuf>,
-) -> Result<u32, AnyError> {
-  let connection = if persistent {
-    let path = state.try_borrow::<LocationDataDir>().ok_or_else(|| {
-      DomExceptionNotSupportedError::new(
-        "LocalStorage is not supported in this context.",
-      )
-    })?;
-    std::fs::create_dir_all(&path.0)?;
-    Connection::open(path.0.join("local_storage"))?
-  } else {
-    Connection::open_in_memory()?
-  };
+) -> Result<&'a Connection, AnyError> {
+  if state.try_borrow::<Connection>().is_none() {
+    let conn = if persistent {
+      let path = state.try_borrow::<LocationDataDir>().ok_or_else(|| {
+        DomExceptionNotSupportedError::new(
+          "LocalStorage is not supported in this context.",
+        )
+      })?;
+      std::fs::create_dir_all(&path.0)?;
+      Connection::open(path.0.join("local_storage"))?
+    } else {
+      Connection::open_in_memory()?
+    };
 
-  connection.execute(
-    "CREATE TABLE IF NOT EXISTS data (key VARCHAR UNIQUE, value VARCHAR)",
-    params![],
-  )?;
+    conn.execute(
+      "CREATE TABLE IF NOT EXISTS data (key VARCHAR UNIQUE, value VARCHAR)",
+      params![],
+    )?;
 
-  let rid = state
-    .resource_table
-    .add(WebStorageConnectionResource(connection));
-  Ok(rid)
+    state.put(conn);
+  }
+
+  Ok(state.borrow::<Connection>())
 }
 
 pub fn op_webstorage_length(
   state: &mut OpState,
-  rid: u32,
-  _zero_copy: Option<ZeroCopyBuf>,
+  persistent: bool,
+  _: (),
 ) -> Result<u32, AnyError> {
-  let resource = state
-    .resource_table
-    .get::<WebStorageConnectionResource>(rid)
-    .ok_or_else(bad_resource_id)?;
+  let conn = get_webstorage(state, persistent)?;
 
-  let mut stmt = resource.0.prepare("SELECT COUNT(*) FROM data")?;
+  let mut stmt = conn.prepare("SELECT COUNT(*) FROM data")?;
 
   let length: u32 = stmt.query_row(params![], |row| row.get(0))?;
 
   Ok(length)
 }
 
-#[derive(Deserialize)]
-#[serde(rename_all = "camelCase")]
-pub struct KeyArgs {
-  rid: u32,
-  index: u32,
-}
-
 pub fn op_webstorage_key(
   state: &mut OpState,
-  args: KeyArgs,
-  _zero_copy: Option<ZeroCopyBuf>,
+  index: u32,
+  persistent: bool,
 ) -> Result<Option<String>, AnyError> {
-  let resource = state
-    .resource_table
-    .get::<WebStorageConnectionResource>(args.rid)
-    .ok_or_else(bad_resource_id)?;
+  let conn = get_webstorage(state, persistent)?;
 
-  let mut stmt = resource
-    .0
-    .prepare("SELECT key FROM data LIMIT 1 OFFSET ?")?;
+  let mut stmt = conn.prepare("SELECT key FROM data LIMIT 1 OFFSET ?")?;
 
   let key: Option<String> = stmt
-    .query_row(params![args.index], |row| row.get(0))
+    .query_row(params![index], |row| row.get(0))
     .optional()?;
 
   Ok(key)
@@ -135,7 +107,6 @@ pub fn op_webstorage_key(
 #[derive(Deserialize)]
 #[serde(rename_all = "camelCase")]
 pub struct SetArgs {
-  rid: u32,
   key_name: String,
   key_value: String,
 }
@@ -143,16 +114,12 @@ pub struct SetArgs {
 pub fn op_webstorage_set(
   state: &mut OpState,
   args: SetArgs,
-  _zero_copy: Option<ZeroCopyBuf>,
+  persistent: bool,
 ) -> Result<(), AnyError> {
-  let resource = state
-    .resource_table
-    .get::<WebStorageConnectionResource>(args.rid)
-    .ok_or_else(bad_resource_id)?;
+  let conn = get_webstorage(state, persistent)?;
 
-  let mut stmt = resource
-    .0
-    .prepare("SELECT SUM(pgsize) FROM dbstat WHERE name = 'data'")?;
+  let mut stmt =
+    conn.prepare("SELECT SUM(pgsize) FROM dbstat WHERE name = 'data'")?;
   let size: u32 = stmt.query_row(params![], |row| row.get(0))?;
 
   if size >= 5000000 {
@@ -162,7 +129,7 @@ pub fn op_webstorage_set(
     );
   }
 
-  resource.0.execute(
+  conn.execute(
     "INSERT OR REPLACE INTO data (key, value) VALUES (?, ?)",
     params![args.key_name, args.key_value],
   )?;
@@ -170,68 +137,43 @@ pub fn op_webstorage_set(
   Ok(())
 }
 
-#[derive(Deserialize)]
-#[serde(rename_all = "camelCase")]
-pub struct GetArgs {
-  rid: u32,
-  key_name: String,
-}
-
 pub fn op_webstorage_get(
   state: &mut OpState,
-  args: GetArgs,
-  _zero_copy: Option<ZeroCopyBuf>,
+  key_name: String,
+  persistent: bool,
 ) -> Result<Option<String>, AnyError> {
-  let resource = state
-    .resource_table
-    .get::<WebStorageConnectionResource>(args.rid)
-    .ok_or_else(bad_resource_id)?;
+  let conn = get_webstorage(state, persistent)?;
 
-  let mut stmt = resource.0.prepare("SELECT value FROM data WHERE key = ?")?;
+  let mut stmt = conn.prepare("SELECT value FROM data WHERE key = ?")?;
 
   let val = stmt
-    .query_row(params![args.key_name], |row| row.get(0))
+    .query_row(params![key_name], |row| row.get(0))
     .optional()?;
 
   Ok(val)
 }
 
-#[derive(Deserialize)]
-#[serde(rename_all = "camelCase")]
-pub struct RemoveArgs {
-  rid: u32,
-  key_name: String,
-}
-
 pub fn op_webstorage_remove(
   state: &mut OpState,
-  args: RemoveArgs,
-  _zero_copy: Option<ZeroCopyBuf>,
+  key_name: String,
+  persistent: bool,
 ) -> Result<(), AnyError> {
-  let resource = state
-    .resource_table
-    .get::<WebStorageConnectionResource>(args.rid)
-    .ok_or_else(bad_resource_id)?;
+  let conn = get_webstorage(state, persistent)?;
 
-  resource
-    .0
-    .execute("DELETE FROM data WHERE key = ?", params![args.key_name])?;
+  conn.execute("DELETE FROM data WHERE key = ?", params![key_name])?;
 
   Ok(())
 }
 
 pub fn op_webstorage_clear(
   state: &mut OpState,
-  rid: u32,
-  _zero_copy: Option<ZeroCopyBuf>,
+  persistent: bool,
+  _: (),
 ) -> Result<(), AnyError> {
-  let resource = state
-    .resource_table
-    .get::<WebStorageConnectionResource>(rid)
-    .ok_or_else(bad_resource_id)?;
+  let conn = get_webstorage(state, persistent)?;
 
-  resource.0.execute("DROP TABLE data", params![])?;
-  resource.0.execute(
+  conn.execute("DROP TABLE data", params![])?;
+  conn.execute(
     "CREATE TABLE data (key VARCHAR UNIQUE, value VARCHAR)",
     params![],
   )?;
@@ -241,15 +183,12 @@ pub fn op_webstorage_clear(
 
 pub fn op_webstorage_iterate_keys(
   state: &mut OpState,
-  rid: u32,
-  _zero_copy: Option<ZeroCopyBuf>,
+  persistent: bool,
+  _: (),
 ) -> Result<Vec<String>, AnyError> {
-  let resource = state
-    .resource_table
-    .get::<WebStorageConnectionResource>(rid)
-    .ok_or_else(bad_resource_id)?;
+  let conn = get_webstorage(state, persistent)?;
 
-  let mut stmt = resource.0.prepare("SELECT key FROM data")?;
+  let mut stmt = conn.prepare("SELECT key FROM data")?;
 
   let keys = stmt
     .query_map(params![], |row| row.get::<_, String>(0))?

--- a/extensions/webstorage/lib.rs
+++ b/extensions/webstorage/lib.rs
@@ -1,4 +1,4 @@
-// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
 
 use deno_core::error::AnyError;
 use deno_core::include_js_files;

